tensorflow/stream_executor/rocm/rocm_blas.cc

```diff
@@ -1811,16 +1811,61 @@ bool ROCMBlas::DoBlasGemmWithAlgorithm(
   return false;
 }
 
-template <typename T>
-struct EigenHalfToRocBlasHalf {
-  using type = T;
-};
-
 template <>
-struct EigenHalfToRocBlasHalf<Eigen::half> {
+struct ROCMBlas::EigenHalfToRocBlasHalf<Eigen::half> {
   using type = rocblas_half;
 };
 
+template <typename T>
+port::Status ROCMBlas::AllocateStridedBuffer(
+    const std::vector<typename EigenHalfToRocBlasHalf<T>::type *>& raw_ptrs, int batch_count, 
+    uint64_t batch_stride, ScratchAllocator* scratch_allocator, Stream * stream, 
+    std::unique_ptr<TemporaryDeviceMemory<typename EigenHalfToRocBlasHalf<T>::type>>& temp_memory,
+    DeviceMemory<typename EigenHalfToRocBlasHalf<T>::type>& device_memory) {
+  using MAPPED_T = typename EigenHalfToRocBlasHalf<T>::type;
+
+  bool needs_allocate_strided = false;
+  for (int i = 1; i < batch_count; ++i) {
+    uint64_t tmp_batch_stride = raw_ptrs[i] - raw_ptrs[i - 1];
+    if (tmp_batch_stride != batch_stride) {
+      needs_allocate_strided = true;
+      break;
+    }
+  }
+
+  size_t matrix_byte_size = batch_stride * sizeof(MAPPED_T);
+  size_t matrix_batch_byte_size = matrix_byte_size * batch_count;
+
+  // No need to do re-allocation, take the short cut and return
+  if (!needs_allocate_strided) {
+    device_memory = DeviceMemory<MAPPED_T>(DeviceMemoryBase(raw_ptrs[0], matrix_batch_byte_size));
+    return port::Status::OK();
+  }
+
+  if (scratch_allocator != nullptr){
+    SE_ASSIGN_OR_RETURN(DeviceMemory<uint8> batch_matrix_bytes,
+                        scratch_allocator->AllocateBytes(stream, matrix_batch_byte_size));
+    device_memory = DeviceMemory<MAPPED_T>(batch_matrix_bytes);
+  } else {
+    SE_ASSIGN_OR_RETURN(temp_memory,
+                        stream->AllocateTemporaryArray<MAPPED_T>(matrix_batch_byte_size));
+    device_memory = DeviceMemory<MAPPED_T>(*temp_memory->mutable_device_memory());
+  }
+
+  for (int i = 0; i < batch_count; ++i) {
+    char * device_memory_ptr = static_cast<char *>(device_memory.opaque());
+    DeviceMemoryBase src_mem = DeviceMemoryBase(raw_ptrs[i], matrix_byte_size); 
+    DeviceMemoryBase target_mem = DeviceMemoryBase(device_memory_ptr + i * matrix_byte_size, matrix_byte_size);
+    bool a_status = stream->ThenMemcpy(&target_mem, src_mem, matrix_byte_size).ok();
+    //parent_->Deallocate(&src_mem);
```

Comment:
I'm confused at how to de-allocate src_mem at this point. After properly copying all src_mem to target_mem, the src_mem should be cleaned up to avoid a memory leak. However, if I do explicit clean up using `parent_->Deallocate()`, using GpuExecutor, the program crashes. 

A further investigation shows that the src_mem originate from the in_x, [here](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/core/kernels/batch_matmul_op_impl.h#L297) in batch_matmul_op_impl. Both in_x, and in_y come from the BaseBatchMatMulOp's [Compute() method's context](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/core/kernels/batch_matmul_op_impl.h#L595). I have trouble figuring out the ownership of the context's owner, therefore not having much clue of how to properly free it. 

@deven-amd, could you help out here?

Reply:
I don't think there would be a leak, and we shouldn't explicitly deallocate them. `in_x`, `in_y` are read only, their associated memory are managed by TF runtime and would be released when those tensors go out of scope.

Reply:
Good point, thanks for clarifying! Will clean up the comment.